### PR TITLE
Fix b value encoder incorrectly partition long multi-byte string bug

### PR DESCRIPTION
**Pull request by Kenneth-KT**
This fix addresses issue https://github.com/mikel/mail/issues/1019.

### CODE REVIEW

File: lib/mail/encodings.rb [hunk 1]
```diff
@@ -301,5 +301,40 @@ def Encodings.collapse_adjacent_encodings(str)
...
+          new_sum = add_sum
+          last_byte_offset = byte_offset
+        end
+        sum = new_sum
+        byte_offset += chr.bytesize
+      end
+      if byte_offset != last_byte_offset
+        yield (last_byte_offset...byte_offset)
+      end
+    end
```
Comment (jeremy) [inline lib/mail/encodings.rb]: Any ideas on performance impact?

File: lib/mail/version_specific/ruby_1_9.rb [hunk 2]
```diff
@@ -243,6 +243,10 @@ def Ruby19.pick_encoding(charset)
       convert_to_encoding(encoding)
     end
 
+    def Ruby19.string_byteslice(str, range)
+      str.unpack('C*').slice(range).pack('C*').force_encoding(str.encoding)
+    end
+
```
[the remark this answers is not in the document]
Comment (Kenneth-KT) [reply lib/mail/version_specific/ruby_1_9.rb]: Yes, but it would the build with Ruby 1.9.2.

[the remark this answers is not in the document]
Comment (jeremy) [reply lib/mail/version_specific/ruby_1_9.rb]: We use feature detection for that reason. For example:
```ruby
if "string".respond_to?(:byteslice)
  def Ruby19.string_byteslice(str, range)
    str.byteslice(range)
  end
else
  def Ruby19.string_byteslice(str, range)
    …
  end
end
```

[the remark this answers is not in the document]
Comment (Kenneth-KT) [reply lib/mail/version_specific/ruby_1_9.rb]: ok, addressed in latest commit

**Comment by jeremy** [comment]
Thank you @Kenneth-KT! I merged with some refactoring in https://github.com/mikel/mail/commit/ae456e861ba5b2432ab758534c43ee8dc8f61e04 👍

**Pull request closed without merging**